lang/cs/Org.Apache.REEF.Wake/Avro/ProtocolSerializer.cs

```diff
@@ -0,0 +1,175 @@
+                }
+            };
+            deserializeMap.Add(typeof(TMessage).Name, deserialize);
+        }
+
+        /// <summary>
+        /// Serialize the input message and return a byte array.
+        /// </summary>
+        /// <param name="message">An object reference to the messeage to be serialized</param>
+        /// <param name="sequence">A long which cotains the higher level protocols sequence number for the message.</param>
+        /// <returns>A byte array containing the serialized header and message.</returns>
+        public byte[] Write(object message, long sequence) 
+        {
+            Logr.Log(Level.Info, "Serializing message: " + message.GetType().Name);
+            try
+            { 
+                IAvroSerializer<Header> headWriter = AvroSerializer.Create<Header>();
+                using (MemoryStream stream = new MemoryStream())
+                {
+                    Header header = new Header(sequence, message.GetType().Name);
+                    headWriter.Serialize(stream, header);
+
+                    Serialize serialize;
+                    string name = message.GetType().Name;
+                    if (serializeMap.TryGetValue(name, out serialize))
+                    {
+                        serialize(stream, message);
+                    }
+                    else
+                    {
+                        Logr.Log(Level.Info, "Request to seriaiize message of unknown type: " + message.GetType().Name);
+                        throw new Exception("Request to serialize unknown message type.");
+                    }
+                    return stream.GetBuffer();
+                }
+            }
+            catch (Exception e)
+            {
+                Logr.Log(Level.Error, "Failure writing message: " + e.Message);
+                throw e;
+            }
+        }
+
+        /// <summary>
+        /// Read a message from the input byte array.
+        /// </summary>
+        /// <param name="data">The byte array containing a header message and message to be deserialized.</param>
+        /// <param name="observer">An object which implements the IObserver<> interface for the message being deserialized.</param>
+        public void Read(byte[] data, object observer)
+        {
+            try
+            {
+                IAvroSerializer<Header> headReader = AvroSerializer.Create<Header>();
+                using (MemoryStream stream = new MemoryStream(data))
+                {
+                    Header head = headReader.Deserialize(stream);
+                    Deserialize deserialize;
+                    if (deserializeMap.TryGetValue(head.className, out deserialize))
+                    {
+                        deserialize(stream, observer, head.sequence);
+                    }
+                    else
+                    {
+                        throw new Exception("Request to deserialize unknown message type.");
```

Comment:
Use more specific exception class; mention the type in the error message, e.g. do something like
```C#
throw new SerializationException("Request to deserialize unknown message type: " + head.className);
```